app/controllers/autocompletes_controller.rb

```diff
@@ -0,0 +1,16 @@
+class AutocompletesController < ApplicationController
+
+  def index
+    @institution = Institution.find(params[:institution_id])
+    @unique_values = unique_values_for(params[:field_name], params[:query])
+    render json: @unique_values.map { |option| { value: option, label: option } }
+  end
+
+  private
+
+  def unique_values_for(field_name, query)
+    batch_values = check_access(@institution.batches, READ_BATCH).collect(&field_name.to_sym).uniq
+    sample_values = check_access(@institution.samples, READ_BATCH).collect(&field_name.to_sym).uniq
+    (batch_values + sample_values).uniq.compact.select { |value| value.to_s.downcase.include?(query.downcase) }.split(',')
+  end
```

Comment:
There is a typo: you should use `READ_SAMPLE` for samples.

Reply:
We should restrict `params[:field_name]` to a list of acceptable field names to autocomplete. Here, any valid method on batch and Sample will be accepted.

Reply:
We could probably optimize a little to avoid too many allocations, for example collect into a Set instead of building large arrays, uniq, building another array, uniq, etc. **But that's not the worst performance issue**, we have to load all batches and samples into ruby memory (because of entity framework), which probably won't do well with millions of samples, repeated on each character we type on the frontend (:fearful:).

Maybe we can load in batch, to avoid loading everything at once?

```ruby
matcher = Regexp.new(Regexp.escape(query), Regexp::IGNORECASE)
values = Set.new

check_access(@institution.batches, READ_BATCH).select(:core_fields).find_each do |batch|
  value = sample.__send__(field_name)
  values << value if value =~ matcher
end

check_access(@institution.samples, READ_SAMPLE).select(:core_fields).find_each do |sample|
  value = sample.__send__(field_name)
  values << value if value =~ matcher
end
```

Note:
- I'm using a regular expression to avoid `value.downcase` to avoid a String allocation/transform but I didn't benchmark to check which is actually faster.
- I assume that the method receives a symbol for `field_name` so it doesn't have to do cast everywhere.

Reply:
Ideally, we should get #1381 in, so we could do that right in MySQL, but that might do in the meantime.

Your idea of an `autocompletes` table was very good, too.